sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -598,50 +598,44 @@ object NullPropagation extends Rule[LogicalPlan] {
 }
 
 /**
- * Attempts to eliminate reading (unnecessary) NULL values if they are not required for correctness
- * by inserting isNotNull filters in the query plan. These filters are currently inserted beneath
- * existing Filters and Join operators and are inferred based on their data constraints.
+ * Eliminate reading unnecessary values if they are not required for correctness (and can help in
+ * optimizing the query) by inserting relevant filters in the query plan based on an operator's
+ * data constraints. These filters are currently inserted to the existing conditions in the Filter
+ * operators and on either side of Join operators.
  *
  * Note: While this optimization is applicable to all types of join, it primarily benefits Inner and
  * LeftSemi joins.
  */
-object NullFiltering extends Rule[LogicalPlan] with PredicateHelper {
+object InferFiltersFromConstraints extends Rule[LogicalPlan] with PredicateHelper {
+  // We generate a list of additional filters from the operator's existing constraint but remove
+  // those that are either already part of the operator's condition or are part of the operator's
+  // child constraints.
   def apply(plan: LogicalPlan): LogicalPlan = plan transform {
     case filter @ Filter(condition, child) =>
-      // We generate a list of additional isNotNull filters from the operator's existing constraints
-      // but remove those that are either already part of the filter condition or are part of the
-      // operator's child constraints.
-      val newIsNotNullConstraints = filter.constraints.filter(_.isInstanceOf[IsNotNull]) --
+      // For the Filter operator, we try to generate additional filters by only inferring the
+      // IsNotNull constraints. These IsNotNull filters are then used while generating the
+      // physical plan to quickly short circuit the null checks in the generated code.
+      val newFilters = filter.constraints.filter(_.isInstanceOf[IsNotNull]) --
```

Comment:
By the way, slightly unrelated to this discussion but Yin, Davies and I had an offline discussion earlier today and we are considering generating the `isNotNull` filters in the physical plan to prevent some TPCDS regressions due to filter pushdown.